File: claude-agent-sdk-ts/src/pendnt.ts
// Thin pendnt REST client used by demo.ts's canUseTool implementation.
// Mirrors /root/work/plugin/lib/client.mjs (the Claude Code plugin's own helper) and
// the shapes documented in /root/work/app/README.md's "API reference" → "Requests".

export const PENDNT_URL = (process.env.PENDNT_URL ?? "https://api.pendnt.dev").replace(/\/+$/, "");
export const PENDNT_API_KEY = process.env.PENDNT_API_KEY ?? "";

export type RequestStatus = "pending" | "approved" | "denied" | "answered" | "expired" | "cancelled";

export interface PendntRequest {
  id: string;
  kind: "approval" | "question" | "notification";
  title: string;
  details?: string | null;
  options?: string[] | null;
  status: RequestStatus;
  answer?: string | null;
  answered_by?: string | null;
  created_at: string;
  expires_at?: string | null;
  answered_at?: string | null;
  hint?: string;
}

class PendntApiError extends Error {
  status: number;
  body: unknown;
  constructor(status: number, body: unknown, message: string) {
    super(message);
    this.name = "PendntApiError";
    this.status = status;
    this.body = body;
  }
}

async function call<T>(
  method: string,
  path: string,
  body?: unknown,
  { timeoutMs = 30_000 }: { timeoutMs?: number } = {},
): Promise<T> {
  if (!PENDNT_API_KEY) {
    throw new Error("PENDNT_API_KEY is not set");
  }
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(`${PENDNT_URL}${path}`, {
      method,
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${PENDNT_API_KEY}`,
      },
      body: body !== undefined ? JSON.stringify(body) : undefined,
      signal: controller.signal,
    });
    const text = await res.text();
    let json: unknown;
    try {
      json = text ? JSON.parse(text) : {};
    } catch {
      json = { raw: text };
    }
    if (!res.ok) {
      const message =
        (json && typeof json === "object" && "message" in json && typeof (json as { message?: unknown }).message === "string"
          ? (json as { message: string }).message
          : undefined) ?? `${method} ${path} failed: HTTP ${res.status}`;
      throw new PendntApiError(res.status, json, message);
    }
    return json as T;
  } finally {
    clearTimeout(timer);
  }
}

/** POST /v1/requests — kind defaults server-side to "approval". */
export function createRequest(input: {
  kind?: "approval" | "question" | "notification";
  title: string;
  details?: string;
  options?: string[];
  timeout_s?: number;
  wait_s?: number;
}): Promise<PendntRequest> {
  return call<PendntRequest>("POST", "/v1/requests", input);
}

/** GET /v1/requests/:id?wait_s=N — wait_s is clamped to 0-25 server-side. */
export function getRequest(id: string, waitS: number): Promise<PendntRequest> {
  return call<PendntRequest>("GET", `/v1/requests/${encodeURIComponent(id)}?wait_s=${waitS}`, undefined, {
    timeoutMs: (waitS + 10) * 1000,
  });
}

/** POST /v1/notify — fire-and-forget. */
export function notify(message: string, level: "info" | "warning" | "error" = "info"): Promise<{ id: string; status: "answered" }> {
  return call("POST", "/v1/notify", { message, level });
}

/**
 * Create an approval request and long-poll it to a terminal state, honoring a total
 * wall-clock budget (`totalTimeoutS`) built out of repeated 25s server-clamped polls —
 * same pattern as the plugin's PermissionRequest hook
 * (/root/work/plugin/hooks/permission-request.mjs) and examples/curl/approve.sh.
 */
export async function requestApprovalAndWait(
  title: string,
  details: string,
  totalTimeoutS: number,
  signal?: AbortSignal,
): Promise<PendntRequest> {
  const POLL_WAIT_S = 25; // server-enforced max per call
  const startedAt = Date.now();

  let row = await createRequest({
    kind: "approval",
    title,
    details,
    wait_s: POLL_WAIT_S,
    timeout_s: totalTimeoutS,
  });

  while (row.status === "pending") {
    if (signal?.aborted) throw new Error("aborted");
    const elapsedS = (Date.now() - startedAt) / 1000;
    if (elapsedS >= totalTimeoutS) break;
    const waitS = Math.max(1, Math.min(POLL_WAIT_S, Math.ceil(totalTimeoutS - elapsedS)));
    row = await getRequest(row.id, waitS);
  }
  return row;
}

/** True if a resolved request row should be treated as "approved". */
export function isApproved(row: PendntRequest): boolean {
  if (row.status === "approved") return true;
  if (row.status === "answered") {
    // The operator answer page always shows a free-text box, even for approval-kind
    // requests — treat a yes/allow/approve-style answer as approval, fail closed
    // otherwise. Same convention as the Claude Code plugin's PermissionRequest hook.
    const text = (row.answer ?? "").trim().toLowerCase();
    return /^(yes|y|allow|approve|approved|ok|okay)\b/.test(text);
  }
  return false;
}

File: claude-agent-sdk-python/pendnt.py
"""Thin pendnt REST client used by demo.py's can_use_tool implementation.

Mirrors /root/work/plugin/lib/client.mjs (the Claude Code plugin's own helper) and the
shapes documented in /root/work/app/README.md's "API reference" -> "Requests". Uses
only the stdlib (urllib) so this module has zero dependencies beyond the Agent SDK
itself.
"""

from __future__ import annotations

import json
import os
import re
import time
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Any, Optional
from urllib.parse import quote

PENDNT_URL = os.environ.get("PENDNT_URL", "https://api.pendnt.dev").rstrip("/")
PENDNT_API_KEY = os.environ.get("PENDNT_API_KEY", "")

POLL_WAIT_S = 25  # server clamps wait_s to 0-25 regardless of what's sent
_APPROVE_RE = re.compile(r"^(yes|y|allow|approve|approved|ok|okay)\b", re.IGNORECASE)


class PendntApiError(Exception):
    def __init__(self, status: int, body: Any, message: str):
        super().__init__(message)
        self.status = status
        self.body = body


@dataclass
class PendntRequest:
    id: str
    kind: str
    title: str
    status: str
    details: Optional[str] = None
    options: Optional[list[str]] = None
    answer: Optional[str] = None
    answered_by: Optional[str] = None
    created_at: Optional[str] = None
    expires_at: Optional[str] = None
    answered_at: Optional[str] = None
    hint: Optional[str] = None

    @classmethod
    def from_json(cls, data: dict[str, Any]) -> "PendntRequest":
        known = {f for f in cls.__dataclass_fields__}
        return cls(**{k: v for k, v in data.items() if k in known})


def _call(method: str, path: str, body: Optional[dict[str, Any]] = None, timeout_s: float = 30.0) -> dict[str, Any]:
    if not PENDNT_API_KEY:
        raise RuntimeError("PENDNT_API_KEY is not set")
    url = f"{PENDNT_URL}{path}"
    data = json.dumps(body).encode("utf-8") if body is not None else None
    req = urllib.request.Request(
        url,
        data=data,
        method=method,
        headers={
            "content-type": "application/json",
            "authorization": f"Bearer {PENDNT_API_KEY}",
        },
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout_s) as resp:
            text = resp.read().decode("utf-8")
            return json.loads(text) if text else {}
    except urllib.error.HTTPError as e:
        text = e.read().decode("utf-8", errors="replace")
        try:
            parsed = json.loads(text) if text else {}
        except json.JSONDecodeError:
            parsed = {"raw": text}
        message = parsed.get("message") if isinstance(parsed, dict) else None
        raise PendntApiError(e.code, parsed, message or f"{method} {path} failed: HTTP {e.code}") from e


def create_request(
    title: str,
    *,
    kind: str = "approval",
    details: str = "",
    options: Optional[list[str]] = None,
    timeout_s: int = 86400,
    wait_s: int = 0,
) -> PendntRequest:
    """POST /v1/requests"""
    body: dict[str, Any] = {"kind": kind, "title": title, "details": details, "timeout_s": timeout_s, "wait_s": wait_s}
    if options is not None:
        body["options"] = options
    return PendntRequest.from_json(_call("POST", "/v1/requests", body, timeout_s=wait_s + 10))


def get_request(request_id: str, wait_s: int = 0) -> PendntRequest:
    """GET /v1/requests/:id?wait_s=N — wait_s is clamped to 0-25 server-side."""
    path = f"/v1/requests/{quote(request_id, safe='')}?wait_s={wait_s}"
    return PendntRequest.from_json(_call("GET", path, timeout_s=wait_s + 10))


def notify(message: str, level: str = "info") -> dict[str, Any]:
    """POST /v1/notify — fire-and-forget."""
    return _call("POST", "/v1/notify", {"message": message, "level": level})


def request_approval_and_wait(title: str, details: str, total_timeout_s: float) -> PendntRequest:
    """Create an approval request and long-poll it to a terminal state, honoring a
    total wall-clock budget built out of repeated 25s server-clamped polls — same
    pattern as the plugin's PermissionRequest hook
    (/root/work/plugin/hooks/permission-request.mjs) and examples/curl/approve.sh.
    """
    started_at = time.monotonic()
    row = create_request(title, details=details, wait_s=POLL_WAIT_S, timeout_s=int(total_timeout_s))

    while row.status == "pending":
        elapsed_s = time.monotonic() - started_at
        if elapsed_s >= total_timeout_s:
            break
        wait_s = max(1, min(POLL_WAIT_S, int(total_timeout_s - elapsed_s) + 1))
        row = get_request(row.id, wait_s=wait_s)

    return row


def is_approved(row: PendntRequest) -> bool:
    """True if a resolved request row should be treated as 'approved'."""
    if row.status == "approved":
        return True
    if row.status == "answered":
        # The operator answer page always shows a free-text box, even for
        # approval-kind requests -- treat a yes/allow/approve-style answer as
        # approval, fail closed otherwise. Same convention as the Claude Code
        # plugin's PermissionRequest hook.
        text = (row.answer or "").strip()
        return bool(_APPROVE_RE.match(text))
    return False
